src/routers/alpha-router/functions/get-candidate-pools.ts

```diff
@@ -197,13 +199,140 @@ class SubcategorySelectionPools<SubgraphPool> {
   constructor(
     public pools: SubgraphPool[],
     public readonly poolsNeeded: number
-  ) {}
+  ) {
+  }
 
   public hasEnoughPools(): boolean {
     return this.pools.length >= this.poolsNeeded;
   }
 }
 
+export type CrossLiquidityCandidatePools = {
+  v2Pools: V2SubgraphPool[];
+  v3Pools: V3SubgraphPool[];
+}
+
+/**
+ * Function that finds any missing pools that were not selected by the heuristic but that would
+ *   create a route with the topPool by TVL with either tokenIn or tokenOut across protocols.
+ *
+ *   e.g. In V2CandidatePools we found that wstETH/DOG is the most liquid pool,
+ *        then in V3CandidatePools ETH/wstETH is *not* the most liquid pool, so it is not selected
+ *        This process will look for that pool in order to complete the route.
+ *
+ */
+export async function getMixedCrossLiquidityCandidatePools({
+  tokenIn,
+  tokenOut,
+  blockNumber,
+  v2SubgraphProvider,
+  v3SubgraphProvider,
+  v2Candidates,
+  v3Candidates,
+}: MixedCrossLiquidityCandidatePoolsParams): Promise<CrossLiquidityCandidatePools> {
+  const v2Pools = (await v2SubgraphProvider.getPools(tokenIn, tokenOut, {
+    blockNumber,
+  })).sort((a, b) => b.reserve - a.reserve);
+  const v3Pools = (await v3SubgraphProvider.getPools(tokenIn, tokenOut, {
+    blockNumber,
+  })).sort((a, b) => b.tvlUSD - a.tvlUSD);
```

Comment:
looks like we will re-download the subgraph pools here. Even though we cache the pools into s3 buckets, re-downloading the pools, especially v2 pools, can add second worth of tail latency on the endpoint right?